src/Microsoft.ML.FastTree/Training/EarlyStoppingCriteria.cs

```diff
@@ -35,24 +41,30 @@ public abstract class IEarlyStoppingCriterion
         /// <param name="isBestCandidate">True if the current result is the best ever.</param>
         /// <returns>If true, the learning should stop.</returns>
         public abstract bool CheckScore(float validationScore, float trainingScore, out bool isBestCandidate);
+
+        /// <summary>
+        /// Create <see cref="IEarlyStoppingCriterionFactory"/> for supporting legacy infra built upon <see cref="IComponentFactory"/>.
+        /// </summary>
+        internal abstract IEarlyStoppingCriterionFactory BuildFactory();
```

Comment:
>internal abstract IEarlyStoppingCriterionFactory BuildFactory(); [](start = 8, length = 64)

How would 3rd party devs create their own EarlyStoppingRuleBase implementations to supply into FastTree algos? This abstract internal method will prevent that

Reply:
Component authorship is not a goal for v1, even at the level of people being able to create their own `ITransformer` implementations. This is fine. Inf act I would go even further by making this thing have a `private protected` constructor.

---
In reply to: [262757300](https://github.com/dotnet/machinelearning/pull/2851#discussion_r262757300) [](ancestors = 262757300)

Reply:
Sounds good Tom. I also think creating a proper stopping rule is a challenge to many C# developers.

---
In reply to: [263023672](https://github.com/dotnet/machinelearning/pull/2851#discussion_r263023672) [](ancestors = 263023672,262757300)

Reply:
In this case if you have only restricted set of EarlyStoppingRule - why dont you go with enum? I dont understand creating such a complicated code where enum will suffice

---
In reply to: [263048215](https://github.com/dotnet/machinelearning/pull/2851#discussion_r263048215) [](ancestors = 263048215,263023672,262757300)

Reply:
I have explained that this can't be enum in the original issue (#2520). Stopping rule itself can contain states, parameters, and so on; it's more than an enum.

---
In reply to: [263057699](https://github.com/dotnet/machinelearning/pull/2851#discussion_r263057699) [](ancestors = 263057699,263048215,263023672,262757300)

Reply:
I would like to discuss this further. You are exposing EarlyStoppingRule and dont allow devs to use it. If you would decouple Factory from Class/Interface you would get a smaller code change and will allow 3rd party devs create their own stopping rule - same as I did for LossFunction

---
In reply to: [263057699](https://github.com/dotnet/machinelearning/pull/2851#discussion_r263057699) [](ancestors = 263057699,263048215,263023672,262757300)

Reply:
Just make a note for our internal discussion --- public area should be as small as possible while all vital functions are still maintained. Thus, we don't allow the implementation of this base class.

---
In reply to: [263063854](https://github.com/dotnet/machinelearning/pull/2851#discussion_r263063854) [](ancestors = 263063854,263057699,263048215,263023672,262757300)

Reply:
I am ok as far as this pattern is an exception and doesnt propagate to things like LossFunction

---
In reply to: [263067053](https://github.com/dotnet/machinelearning/pull/2851#discussion_r263067053) [](ancestors = 263067053,263063854,263057699,263048215,263023672,262757300)

Reply:
This may not affect LossFunction because it's an interface. If LossFunction is a class, the same pattern may apply. As Tom and Ivan have mentioned somewhere, public area should be as small as possible. On the other hand, another pattern here may affect LossFunction where we have a property to hide a field.

---
In reply to: [263068114](https://github.com/dotnet/machinelearning/pull/2851#discussion_r263068114) [](ancestors = 263068114,263067053,263063854,263057699,263048215,263023672,262757300)